packages/react/src/components/Comments/FloatingThreadController.tsx

```diff
@@ -0,0 +1,108 @@
+import {
+  BlockSchema,
+  DefaultBlockSchema,
+  DefaultInlineContentSchema,
+  DefaultStyleSchema,
+  InlineContentSchema,
+  StyleSchema,
+} from "@blocknote/core";
+import { UseFloatingOptions, flip, offset, shift } from "@floating-ui/react";
+import {
+  ComponentProps,
+  FC,
+  useCallback,
+  useEffect,
+  useLayoutEffect,
+} from "react";
+
+import { useBlockNoteEditor } from "../../hooks/useBlockNoteEditor.js";
+import { useUIElementPositioning } from "../../hooks/useUIElementPositioning.js";
+import { useUIPluginState } from "../../hooks/useUIPluginState.js";
+import { Thread } from "./Thread.js";
+
+/**
+ * This component is used to display a thread in a floating card.
+ * It can be used when the user clicks on a thread / comment in the document.
+ */
+export const FloatingThreadController = <
+  B extends BlockSchema = DefaultBlockSchema,
+  I extends InlineContentSchema = DefaultInlineContentSchema,
+  S extends StyleSchema = DefaultStyleSchema
+>(props: {
+  floatingThread?: FC<ComponentProps<typeof Thread>>;
+  floatingOptions?: Partial<UseFloatingOptions>;
+}) => {
+  const editor = useBlockNoteEditor<B, I, S>();
+
+  if (!editor.comments) {
+    throw new Error(
+      "FloatingComposerController can only be used when BlockNote editor has enabled comments"
+    );
+  }
+
+  const state = useUIPluginState(
+    editor.comments.onUpdate.bind(editor.comments)
+  );
+
+  const { isMounted, ref, style, getFloatingProps, setReference } =
+    useUIElementPositioning(!!state?.selectedThreadId, null, 5000, {
+      placement: "bottom",
+      middleware: [offset(10), shift(), flip()],
+      onOpenChange: (open) => {
+        if (!open) {
+          editor.comments?.selectThread(undefined);
+          editor.focus();
+        }
+      },
+      ...props.floatingOptions,
+    });
+
+  // Positioning with [data-bn-thread-id] attribute is a bit hacky,
+  // we could probably also use the thread position from the plugin state?
+  // for now, this works ok
+  const updateRef = useCallback(() => {
+    if (!state?.selectedThreadId) {
+      return;
+    }
+
+    const el = editor.domElement?.querySelector(
+      `[data-bn-thread-id="${state?.selectedThreadId}"]`
+    );
+    if (el) {
+      setReference(el);
+    }
+  }, [setReference, editor, state?.selectedThreadId]);
```

Comment:
Definitely strange, I don't quite understand how this all works

Reply:
@YousefED can you explain what's happening here? Seems a bit hacky to use `querySelector`